### PR TITLE
initial implementation of signed message create and decode functions

### PR DESCRIPTION
**Pull request by elv-serban**
matching authd PR is https://github.com/qluvio/elv-master/pull/1128

### CODE REVIEW

File: src/ElvClient.js [hunk 1]
```diff
@@ -975,6 +975,75 @@ class ElvClient {
...
+            const msgBytes = Utils.FromB58(signedMessage.slice(4));
+            const signature = msgBytes.slice(0, 65);
+            const msg = msgBytes.slice(65);
+            const obj = JSON.parse(msg);
+
+            // TODO - Verify signature and extract signer
+            //
+            signerAddr = null;
+
+            res = {
```
Comment (elv-kevin) [inline src/ElvClient.js:1033]: makes more sense to just return here I think

File: src/ElvClient.js [hunk 2]
```diff
@@ -975,6 +975,75 @@ class ElvClient {
...
+                type: type,
+                obj: obj,
```
Comment (elv-kevin) [inline src/ElvClient.js:1035]: return value json should be called `message`, and let's also call addr `address`, more consistent with the rest of the client

**Review by elv-kevin — APPROVED** [review]
otherwise looks good

File: src/ElvClient.js [hunk 3]
```diff
@@ -975,6 +975,75 @@ class ElvClient {
...
+      ])
+    )}`;
+  }
+
+  /**
+   * Verify and decode a signed message (JSON).
+   *
+   * @methodGroup Authorization
+   * @namedParams
+   * @param {string} signedMessage - a signed message as created by CreateSignedMessageJSON
```
Comment (elv-kevin) [inline src/ElvClient.js:1015]: Add a return doc to this method as well, something like
`* @returns {Promise<Object>} - The decoded message, signer address, signature and signature type`

**Review by elv-todd — APPROVED** [review]

**Comment by elv-preethi** [comment]
added signer to `DecodeSignedMessageJSON` and test in `testScripts/CreateAndDecodeSignedMessage.js`

```
node testScripts/CreateAndDecodeSignedMessage.js --msg "{"test":"msg"}" --config-url https://demov3.net955210.contentfabric.io/config
mje_2Pfankacc1Cjsvtc1CAU3wxZduziDf5KmjwXiWYnNSo1KvGRJR6Y6XBZdcWQUyqZfyhYhabwTA513Zsa6tEX6mWeE6A2EuJNtb9sMt12yX
{
  type: 'mje_',
  obj: '{test:msg}',
  addr: '0xBF01001eaF0dE9B7b88BCD2DC0f0577A768Efd19',
  signature: '0xbd3790b4d46277d8903064b919c859049e3c65b176dd454d51f1b6a6bd26b432190e1b91a258dc053f686853f6699585ad979056de3c7a86bf3fc12bc83e9e001c'
}
```

File: src/ElvClient.js (hunk 1 above)
Comment (elv-preethi) [reply src/ElvClient.js]: made the changes.

File: src/ElvClient.js (hunk 2 above)
Comment (elv-preethi) [reply src/ElvClient.js:1034]: In authd we have this:
```
type DecodedMessage struct {
	Type      string                 `json:"type"`
	Obj       map[string]interface{} `json:"obj"`
	Signature string                 `json:"signature"`
}
```

Also, signer not included.

File: src/ElvClient.js [hunk 4]
```diff
@@ -975,6 +975,71 @@ class ElvClient {
...
+            const msg = msgBytes.slice(65);
+            const obj = JSON.parse(msg);
+
+            const prefixedMsgHash = Ethers.utils.keccak256(Buffer.from(`\x19Ethereum Signed Message:\n${msg.length}${msg}`, "utf-8"));
+            const signerAddr = Ethers.utils.recoverAddress(prefixedMsgHash, signature);
+
+            return {
+                type: type,
+                message: obj,
+                address: signerAddr,
```
Comment (elv-todd) [inline src/ElvClient.js:1035]: should this be signer_address?

File: src/ElvClient.js (hunk 2 above)
Comment (elv-todd) [reply src/ElvClient.js]: change it in the authd PR, branch `add-nft-claim-entitlement-action`

it also needs to be updated in
- media-wallet PR
- elv-sample-custom-app-purchase

but we should just finalize it here and pull out the duplicated code in them anyway, so, fine to change formats

File: src/ElvClient.js (hunk 4 above)
Comment (elv-kevin) [reply src/ElvClient.js:1035]: signerAddress is fine

**Pull request merged**